ntoskrnl/se/debug.c

```diff
@@ -0,0 +1,388 @@
+ *
+ * @return
+ * Returns a converted ACE type strings. If no
+ * known ACE type is found, it will return
+ * UNKNOWN TYPE.
+ */
+static
+PCSTR
+SepGetAceTypeString(
+    _In_ UCHAR AceType)
+{
+#define TOSTR(x)    #x
+    static const PCSTR AceTypes[] =
+    {
+        TOSTR(ACCESS_ALLOWED_ACE_TYPE),
+        TOSTR(ACCESS_DENIED_ACE_TYPE),
+        TOSTR(SYSTEM_AUDIT_ACE_TYPE),
+        TOSTR(SYSTEM_ALARM_ACE_TYPE),
+        TOSTR(ACCESS_ALLOWED_COMPOUND_ACE_TYPE),
+        TOSTR(ACCESS_ALLOWED_OBJECT_ACE_TYPE),
+        TOSTR(ACCESS_DENIED_OBJECT_ACE_TYPE),
+        TOSTR(SYSTEM_AUDIT_OBJECT_ACE_TYPE),
+        TOSTR(SYSTEM_ALARM_OBJECT_ACE_TYPE),
+        TOSTR(ACCESS_ALLOWED_CALLBACK_ACE_TYPE),
+        TOSTR(ACCESS_DENIED_CALLBACK_ACE_TYPE),
+        TOSTR(ACCESS_ALLOWED_CALLBACK_OBJECT_ACE_TYPE),
+        TOSTR(ACCESS_DENIED_CALLBACK_OBJECT_ACE_TYPE),
+        TOSTR(SYSTEM_AUDIT_CALLBACK_ACE_TYPE),
+        TOSTR(SYSTEM_ALARM_CALLBACK_ACE_TYPE),
+        TOSTR(SYSTEM_AUDIT_CALLBACK_OBJECT_ACE_TYPE),
+        TOSTR(SYSTEM_ALARM_CALLBACK_OBJECT_ACE_TYPE),
+        TOSTR(SYSTEM_MANDATORY_LABEL_ACE_TYPE),
+    };
+#undef TOSTR
+
+    if (AceType < RTL_NUMBER_OF(AceTypes))
+        return AceTypes[AceType];
+    else
+        return "UNKNOWN TYPE";
+}
+
+/**
+ * @brief
+ * Dumps the ACE flags to the debugger output.
+ */
+static
+VOID
+SepDumpAceFlags(
+    _In_ UCHAR AceFlags)
+{
+    if (AceFlags & OBJECT_INHERIT_ACE)
+    {
+        DbgPrint("OBJECT_INHERIT_ACE\n");
+    }
+
+    if (AceFlags & CONTAINER_INHERIT_ACE)
+    {
+        DbgPrint("CONTAINER_INHERIT_ACE\n");
+    }
+
+    if (AceFlags & NO_PROPAGATE_INHERIT_ACE)
+    {
+        DbgPrint("NO_PROPAGATE_INHERIT_ACE\n");
+    }
+
+    if (AceFlags & INHERIT_ONLY_ACE)
+    {
+        DbgPrint("INHERIT_ONLY_ACE\n");
+    }
+
+    if (AceFlags & INHERITED_ACE)
+    {
+        DbgPrint("INHERITED_ACE\n");
+    }
+}
+
+/**
+ * @brief
+ * Iterates and dumps each ACE debug info in an ACL.
+ */
+static
+VOID
+SepDumpAces(
+    _In_ PACL Acl)
+{
+    NTSTATUS Status;
+    PACE Ace;
+    ULONG AceNumber;
+    ULONG AceIndex;
+    PSID Sid;
+    UNICODE_STRING SidString;
+
+    /* Loop all ACEs and dump their info */
+    AceNumber = 1;
+    for (AceIndex = 0; AceIndex < Acl->AceCount; AceIndex++)
+    {
+        /* Get the ACE at this index */
+        Status = RtlGetAce(Acl, AceIndex, (PVOID*)&Ace);
+        if (!NT_SUCCESS(Status))
+        {
+            /*
+             * Normally this should never happen.
+             * Just fail gracefully and stop further
+             * debugging of ACEs.
+             */
+            DbgPrint("SepDumpAces(): Failed to find the next ACE, stop dumping info...\n");
+            return;
+        }
+
+        DbgPrint("================== %lu# ACE DUMP INFO ==================\n", AceNumber);
+        DbgPrint("ACE pointer -> %p\n", Ace);
+        DbgPrint("ACE header pointer -> %p\n", Ace->Header);
```

Comment:
```suggestion
        DbgPrint("ACE pointer -> 0x%p\n", Ace);
        DbgPrint("ACE header pointer -> 0x%p\n", Ace->Header);
```

Reply:
Not sure you have to mention "pointer" everytime.
What about displaying the field name instead?

Reply:
Regarding the "What about displaying the field name instead?" suggestion, it's so that, by looking at the dump info, as developers we can directly make it correspond to the structure field in question, without having to look at the dump functions to make the correspondence with the structure field name.